File: js/testapps/agents/src/file-store.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as fs from 'fs';
import { z } from 'genkit';
import { FileSessionStore } from 'genkit/beta';
import * as path from 'path';
import { ai } from './genkit.js';

export { FileSessionStore };

export const fileStore = new FileSessionStore<any, any>('./.snapshots');

export const fileStorePrompt = ai.definePrompt({
  name: 'fileStorePrompt',
  model: 'googleai/gemini-flash-lite-latest',
  input: { schema: z.object({ userName: z.string() }) },
  system: `You are a personal logbook assistant. Always address the user by the name {{ userName }}.`,
});

export const fileStoreAgent = ai.defineSessionFlowFromPrompt({
  promptName: 'fileStorePrompt',
  defaultInput: { userName: 'Stranger' },
  store: fileStore,
});

export const testFileStoreAgent = ai.defineFlow(
  {
    name: 'testFileStoreAgent',
    inputSchema: z.string().default('Alice'),
    outputSchema: z.any(),
  },
  async (userName) => {
    // Run Turn 1
    const turn1 = await fileStoreAgent.run(
      {
        messages: [
          {
            role: 'user',
            content: [
              {
                text: 'Hello! Please log this note: I started studying Genkit today.',
              },
            ],
          },
        ],
      },
      {
        init: {
          state: {
            inputVariables: { userName },
            custom: {},
            messages: [],
            artifacts: [],
          },
        },
      }
    );

    const snapshotId1 = turn1.result.snapshotId!;

    // Now simulate Turn 2 by resuming from the written File snapshot
    const turn2 = await fileStoreAgent.run(
      {
        messages: [
          { role: 'user', content: [{ text: 'What did I study today?' }] },
        ],
      },
      { init: { snapshotId: snapshotId1 } }
    );

    return {
      snapshotId1,
      reply1: turn1.result.message?.content?.map((c) => c.text || '').join(''),
      reply2: turn2.result.message?.content?.map((c) => c.text || '').join(''),
    };
  }
);
export const pruningStore = new FileSessionStore<any, any>(
  './.snapshots-pruning',
  {
    maxPersistedChainLength: 3,
  }
);

export const pruningAgent = ai.defineSessionFlowFromPrompt({
  promptName: 'fileStorePrompt',
  defaultInput: { userName: 'Stranger' },
  store: pruningStore,
});

export const testFileStoreChainPruningAgent = ai.defineFlow(
  {
    name: 'testFileStoreChainPruningAgent',
    inputSchema: z.string().default('Alice'),
    outputSchema: z.any(),
  },
  async (userName) => {
    // Run Turn 1
    const turn1 = await pruningAgent.run(
      {
        messages: [{ role: 'user', content: [{ text: 'Turn 1' }] }],
      },
      {
        init: {
          state: {
            inputVariables: { userName },
            custom: {},
            messages: [],
            artifacts: [],
          },
        },
      }
    );
    const snap1 = turn1.result.snapshotId!;

    // Run Turn 2
    const turn2 = await pruningAgent.run(
      {
        messages: [{ role: 'user', content: [{ text: 'Turn 2' }] }],
      },
      { init: { snapshotId: snap1 } }
    );
    const snap2 = turn2.result.snapshotId!;

    // Run Turn 3
    const turn3 = await pruningAgent.run(
      {
        messages: [{ role: 'user', content: [{ text: 'Turn 3' }] }],
      },
      { init: { snapshotId: snap2 } }
    );
    const snap3 = turn3.result.snapshotId!;

    // Run Turn 4 (Snap 1 should be deleted here since max chain length is 3)
    const turn4 = await pruningAgent.run(
      {
        messages: [{ role: 'user', content: [{ text: 'Turn 4' }] }],
      },
      { init: { snapshotId: snap3 } }
    );

    // Snapshots are stored under <dirPath>/global/<snapshotId>.json
    const snapshotDir = path.join('./.snapshots-pruning', 'global');
    const snap1Exists = fs.existsSync(path.join(snapshotDir, `${snap1}.json`));
    const snap2Exists = fs.existsSync(path.join(snapshotDir, `${snap2}.json`));
    const snap3Exists = fs.existsSync(path.join(snapshotDir, `${snap3}.json`));
    const snap4Exists = fs.existsSync(
      path.join(snapshotDir, `${turn4.result.snapshotId}.json`)
    );

    return {
      snap1Exists,
      snap2Exists,
      snap3Exists,
      snap4Exists,
    };
  }
);

File: py/packages/genkit/src/genkit/_ai/_agents/_session_stores/_file_store.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""File-backed, full-snapshot session store for local development and tests."""

from __future__ import annotations

import asyncio
import contextlib
import os
from typing import Any, Generic

from genkit._ai._agents._session import (
    SessionStore,
    SessionStoreLock,
    SnapshotStatusStream,
    SnapshotSubscriber,
    StateT,
)
from genkit._ai._agents._session_stores._util import (
    SaveFn,
    Subs,
    apply_save,
    assert_safe_snapshot_id,
    iterate_statuses,
    notify,
    require_one_selector,
    select_leaf,
    session_id_of,
    subscribe,
)
from genkit._core._error import GenkitError, RuntimeErrorReason
from genkit._core._typing import SessionSnapshot


class FileSessionStore(SessionStoreLock, SessionStore[StateT], SnapshotSubscriber, Generic[StateT]):
    """File-backed snapshot store: one ``<snapshot_id>.json`` per snapshot."""

    def __init__(
        self,
        directory: str,
        *,
        reject_ambiguous_session: bool = False,
        max_persisted_chain_length: int | None = None,
    ) -> None:
        """Create the store, ensuring ``directory`` exists.

        See :class:`InMemorySessionStore` for ``reject_ambiguous_session``.

        ``max_persisted_chain_length`` caps how many snapshots of a chat's
        history stay on disk: once a chain grows past it, the oldest turns are
        deleted on each save so a long-lived conversation doesn't accumulate
        files forever. Resuming and continuing still work, but you lose the
        ability to rewind or branch past the retained window. Leave it unset to
        keep the full history.
        """
        self.reject_ambiguous = reject_ambiguous_session
        self.max_persisted_chain_length = max_persisted_chain_length
        self.directory = directory
        self.subs: Subs = {}
        os.makedirs(directory, exist_ok=True)

    def path(self, snapshot_id: str) -> str:
        """Return the file path for a snapshot ID."""
        assert_safe_snapshot_id(snapshot_id=snapshot_id)
        return os.path.join(self.directory, f'{snapshot_id}.json')

    def read_sync(self, snapshot_id: str) -> SessionSnapshot | None:
        """Read and validate a snapshot from disk synchronously."""
        path = self.path(snapshot_id)
        if not os.path.exists(path):
            return None
        with open(path, encoding='utf-8') as f:
            return SessionSnapshot.model_validate_json(f.read())

    def write_sync(self, snapshot: SessionSnapshot) -> None:
        """Atomically write a snapshot to disk synchronously."""
        path = self.path(snapshot.snapshot_id)
        temp_path = path + '.tmp'
        with open(temp_path, 'w', encoding='utf-8') as f:
            f.write(snapshot.model_dump_json(indent=2))
        os.replace(temp_path, path)

    def delete_sync(self, snapshot_id: str) -> None:
        """Delete a snapshot file, tolerating one that's already gone."""
        try:
            os.remove(self.path(snapshot_id))
        except FileNotFoundError:
            pass

    def prune_chain_sync(self, leaf: SessionSnapshot) -> None:
        """Trim a chat's ancestry to the newest ``max_persisted_chain_length`` turns.

        Walks ``parent_id`` back from the just-written snapshot and deletes the
        oldest links past the cap. The retained oldest snapshot keeps pointing at
        its (now-deleted) parent, so history reconstruction simply stops at the
        window's edge.
        """
        cap = self.max_persisted_chain_length
        if not cap or cap <= 0:
            return
        chain: list[str] = []
        seen: set[str] = set()
        cur: SessionSnapshot | None = leaf
        # `seen` stops a corrupt/cyclic parent chain from looping forever (each
        # hop is a disk read), the same guard walk_back_to_resumable uses.
        while cur is not None and cur.snapshot_id not in seen:
            seen.add(cur.snapshot_id)
            chain.append(cur.snapshot_id)
            cur = self.read_sync(cur.parent_id) if cur.parent_id else None
        for snapshot_id in chain[cap:]:
            self.delete_sync(snapshot_id)

    def read_session_sync(self, session_id: str) -> list[SessionSnapshot]:
        """Read all snapshots for a session from disk synchronously."""
        out: list[SessionSnapshot] = []
        if not os.path.isdir(self.directory):
            return out
        for name in os.listdir(self.directory):
            if not name.endswith('.json'):
                continue
            try:
                with open(os.path.join(self.directory, name), encoding='utf-8') as f:
                    snap = SessionSnapshot.model_validate_json(f.read())
            except (OSError, ValueError):
                continue
            if session_id_of(snap) == session_id:
                out.append(snap)
        return out

    async def get_snapshot(
        self,
        *,
        snapshot_id: str | None = None,
        session_id: str | None = None,
        context: dict[str, Any] | None = None,
    ) -> SessionSnapshot | None:
        """Return a snapshot by id, or the session's latest leaf, read from disk."""
        require_one_selector(snapshot_id=snapshot_id, session_id=session_id)
        async with self.lock:
            if snapshot_id is not None:
                return await asyncio.to_thread(self.read_sync, snapshot_id)

            assert session_id is not None
            owned = await asyncio.to_thread(self.read_session_sync, session_id)
            return select_leaf(snapshots=owned, session_id=session_id, reject_ambiguous=self.reject_ambiguous)

    async def save_snapshot(
        self,
        snapshot_id: str,
        fn: SaveFn,
        *,
        context: dict[str, Any] | None = None,
    ) -> SessionSnapshot | None:
        """Read-modify-write a snapshot on disk, prune the chain, and notify subscribers."""
        _ = context
        async with self.lock:
            existing = await asyncio.to_thread(self.read_sync, snapshot_id)
            next_snapshot = apply_save(existing=existing, snapshot_id=snapshot_id, fn=fn)
            if next_snapshot is None:
                return None
            if not session_id_of(next_snapshot):
                raise GenkitError(
                    status='INVALID_ARGUMENT',
                    message="FileSessionStore requires 'sessionId' on the snapshot.",
                    reason=RuntimeErrorReason.SESSION_ID_REQUIRED,
                )
            await asyncio.to_thread(self.write_sync, next_snapshot)
            if self.max_persisted_chain_length:
                await asyncio.to_thread(self.prune_chain_sync, next_snapshot)
            notify(subs=self.subs, snapshot_id=next_snapshot.snapshot_id, status=next_snapshot.status)
            return next_snapshot

    async def on_snapshot_status_change(
        self, snapshot_id: str, *, context: dict[str, Any] | None = None
    ) -> SnapshotStatusStream:
        """Return an iterator that yields this snapshot's status changes.

        Call ``aclose()`` (or ``contextlib.aclosing``) to unsubscribe early;
        natural end after a terminal status also unsubscribes.
        """
        async with self.lock:
            current = await asyncio.to_thread(self.read_sync, snapshot_id)
            q = await subscribe(subs=self.subs, snapshot_id=snapshot_id, current=current)

            async def on_close() -> None:
                async with self.lock:
                    qs = self.subs.get(snapshot_id)
                    if qs is None:
                        return
                    with contextlib.suppress(ValueError):
                        qs.remove(q)
                    if not qs:
                        self.subs.pop(snapshot_id, None)

            return iterate_statuses(q, on_close=on_close)
